### ISSUE TITLE
Make buildpack detection errors info level

### ISSUE THREAD

**Issue by jabrown85** [buildpacks/lifecycle#283]
### Summary
When `pack build` is executed no errors are reported in cases where the buildpack's detect phase produced an error. This can occur for many reasons but the most common case is running out of disk space. In this case ALL buildpacks fail with an unexpected error but the error is not displayed via `pack` without  the `--verbose` flag.

---

##### Steps

1. Fill up your docker vm disk space
2. Run `pack build my-image`
3. Observe the build fail with a no detection error with little details on why

##### Current behavior

```
$ pack build riffsquare --builder projectriff/builder:0.6.0-snapshot -e RIFF=1
0.6.0-snapshot: Pulling from projectriff/builder
Digest: sha256:9349cc12f5944982dfe41976acd734a411d87db158ed3c0e818fa3b1fd87a6fb
Status: Image is up to date for projectriff/builder:0.6.0-snapshot
base-cnb: Pulling from cloudfoundry/run
Digest: sha256:c1e5e1fdd95465525df6ecdd99be6f47d37a1e256a22c53287bad08aead92f3f
Status: Image is up to date for cloudfoundry/run:base-cnb
===> DETECTING
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```

##### Expected behavior

```
$ pack build riffsquare --builder projectriff/builder:0.6.0-snapshot -e RIFF=1
0.6.0-snapshot: Pulling from projectriff/builder
Digest: sha256:9349cc12f5944982dfe41976acd734a411d87db158ed3c0e818fa3b1fd87a6fb
Status: Image is up to date for projectriff/builder:0.6.0-snapshot
base-cnb: Pulling from cloudfoundry/run
Digest: sha256:c1e5e1fdd95465525df6ecdd99be6f47d37a1e256a22c53287bad08aead92f3f
Status: Image is up to date for cloudfoundry/run:base-cnb
===> DETECTING
[detector] ======== Error: org.cloudfoundry.openjdk@1.2.14 ========
[detector] mkdir /tmp/plan.422531505: no space left on device
[detector] ======== Error: org.cloudfoundry.buildsystem@1.2.15 ========
[detector] mkdir /tmp/plan.861676380: no space left on device
[detector] ======== Error: io.projectriff.java@0.3.0-BUILD-SNAPSHOT ========
[detector] mkdir /tmp/plan.036661003: no space left on device
...repeat for all buidpacks that produce Errors...
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```
---

##### pack info

`v0.9.0 (git sha: d42c384a39f367588f2653f2a99702db910e5ad7)`

##### docker info

```
Client:
 Debug Mode: false

Server:
 Containers: 3
  Running: 1
  Paused: 0
  Stopped: 2
 Images: 3
 Server Version: 19.03.8
 Storage Driver: overlay2
  Backing Filesystem: <unknown>
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc version: dc9208a3303feef5b3839f4323d9beb36df0a9dd
 init version: fec3683
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.76-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 1.943GiB
 Name: docker-desktop
 ID: WRJQ:2OXJ:SCYH:DMYM:PWZL:U7W3:P7BH:P4BT:4LR5:DGX6:LWYL:CN5W
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 43
  Goroutines: 64
  System Time: 2020-04-03T18:56:25.820913952Z
  EventsListeners: 3
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 Product License: Community Engine

```

**Comment by ekcasey** [comment]
@jabrown85 
A couple questions:

1. Do you think buildpack detection errors should always be info level, or only if zero groups have detected?

2. In the case that zero groups have detected, but they are all `fail` cases rather than `err` cases. What output would you like to see in then?

**Comment by jabrown85** [comment]
Hmm. This is what it looks like today for one of heroku's builders against an empty directory. (using `pack` for simplicity)

```
» pack build this --builder heroku/buildpacks --no-pull
===> DETECTING
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```
verbose:
```
» pack build this --builder heroku/buildpacks --verbose --no-pull
Selected run image heroku/pack:18
Using build cache volume pack-cache-55c498cc1605.build
===> DETECTING
[detector] ======== Output: heroku/ruby@0.0.1 ========
[detector] no
[detector] ======== Results ========
[detector] err:  heroku/ruby@0.0.1 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] err:  heroku/python@0.2 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Output: heroku/jvm@0.1 ========
[detector] JVM
[detector] ======== Output: heroku/maven@0.1 ========
[detector] Could not find a pom.xml file! Please check that it exists and is committed to Git.
[detector] ======== Results ========
[detector] pass: heroku/jvm@0.1
[detector] err:  heroku/maven@0.1 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] err:  heroku/gradle@0.2 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] err:  heroku/scala@0.2 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] err:  heroku/php@0.2 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] err:  heroku/go@0.2 (1)
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] fail: heroku/nodejs-engine@0.4.3
[detector] fail: heroku/nodejs-yarn@0.0.1
[detector] pass: heroku/procfile@0.5
[detector] ======== Results ========
[detector] fail: heroku/nodejs-engine@0.4.3
[detector] fail: heroku/nodejs-npm@0.1.4
[detector] pass: heroku/procfile@0.5
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```

For question 1, I originally thought outputting always on `info` for `err` seems ok. Looking at the output above would that mean it looks something like this though?
```
» pack build this --builder heroku/buildpacks --no-pull
Selected run image heroku/pack:18
Using build cache volume pack-cache-55c498cc1605.build
===> DETECTING
[detector] err:  heroku/ruby@0.0.1 (1)
[detector] err:  heroku/python@0.2 (1)
[detector] err:  heroku/maven@0.1 (1)
[detector] err:  heroku/gradle@0.2 (1)
[detector] err:  heroku/scala@0.2 (1)
[detector] err:  heroku/php@0.2 (1)
[detector] err:  heroku/go@0.2 (1)
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```

In this case we've got a few buildpacks doing `fail` compared to `err` on detection. I think the ones doing `err` are old buildpacks being wrapped in a CNB shim. The problem with this output is it implies you need to fix one of the `err` buildpacks but it omits `fail` which might be what you really want to fix. In the case above there is a nodejs buildpack that is `fail` which is what I really want to see in this case.

It would look even more confusing if this was a passing build.

```
» pack build this --builder heroku/buildpacks --no-pull
Selected run image heroku/pack:18
Using build cache volume pack-cache-55c498cc1605.build
===> DETECTING
[detector] err:  heroku/ruby@0.0.1 (1)
[detector] err:  heroku/python@0.2 (1)
[detector] err:  heroku/maven@0.1 (1)
[detector] err:  heroku/gradle@0.2 (1)
[detector] err:  heroku/scala@0.2 (1)
[detector] err:  heroku/php@0.2 (1)
[detector] heroku/go       0.2
[detector] heroku/procfile 0.5
===> ANALYZING
[analyzer] Previous image with name "index.docker.io/library/this:latest" not found
===> RESTORING
===> BUILDING
[builder] -----> Fetching jq... done
[builder] -----> Fetching stdlib.sh.v8... done
[builder] ----->
[builder]        Detected go modules via go.mod
[builder] ----->
[builder]        Detected Module Name: github.com/heroku/go-getting-started
[builder] ----->
[builder]  !!    The go.mod file for this project does not specify a Go version
[builder]  !!
[builder]  !!    Defaulting to go1.12.17
[builder]  !!
[builder]  !!    For more details see: https://devcenter.heroku.com/articles/go-apps-with-modules#build-configuration
[builder]  !!
[builder] -----> New Go Version, clearing old cache
[builder] -----> Installing go1.12.17
....
```

Regardless of the above...

Question 1: I think `err` should always show regardless of detections overall success. The shim behavior here seems unexpected and we should fix that.

Question 2: I think showing all the `fail` buildpacks could be ok if detect failed. I don't feel strongly about this though. The current behavior is pretty decent. Outputting 30 buildpack failures might not be better than no output at all (in the case of `pack` at least)

```
» pack build this --builder heroku/buildpacks --no-pull
Selected run image heroku/pack:18
Using build cache volume pack-cache-55c498cc1605.build
===> DETECTING
[detector] fail: Group 1
[detector] fail: heroku/nodejs-engine@0.4.3
[detector] fail: heroku/nodejs-yarn@0.0.1
[detector] pass: heroku/procfile@0.5
[detector] fail: Group 2
[detector] fail: heroku/nodejs-engine@0.4.3
[detector] fail: heroku/nodejs-npm@0.1.4
[detector] pass: heroku/procfile@0.5
....
[detector] ERROR: No buildpack groups passed detection.
[detector] ERROR: Please check that you are running against the correct path.
[detector] ERROR: failed to detect: no buildpacks participating
ERROR: failed with status code: 6
```

**Comment by ekcasey** [comment]
@jabrown85 Yeah, it looks like the shimmed Heroku buildpack's are not emitting exit code `100` on failure so it is interpreted as an error rather than a failure by the lifecycle, per [this section](https://github.com/buildpacks/spec/blob/master/buildpack.md#detection) of the buildpack API spec.

As you suggested, I think we can go with: 
1. Info level for all errors*
1. Keep the current log level behavior for failing buildpacks

* follow up: I am assuming we want both the fact that it errored and the buildpack's output in this case?

The above is easy compared to adding info level failures only in the case when all groups fail to detect (which requires larger code changes), and we can always change that later if we continue to get constructive feedback about detect logs.

**Comment by jabrown85** [comment]
followup: I think it makes sense to have both.

**Comment by dfreilich** [comment]
@ekcasey Has there been any movement on this? Should someone pick it up?

**Comment by ekcasey** [comment]
Resolved by #351
